keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1107 @@
+  "serviceAccountAnnotations": {
+    "domain.io/identity-id": "12345",
+    "domain.io/identity-type": "user"
+  }
+}
+```
+
+#### Caching Credentials
+
+The cache key generation will be updated to support new caching modes while preserving the existing behavior when the audience field is not set.
+
+##### Existing behavior
+
+If the `serviceAccountTokenAudience` field is not set in the provider configuration for the plugin, the key generation should fall back to the current behavior: using only the image registry URL or other existing identifiers.
+
+##### New behavior when the `serviceAccountTokenAudience` field is set
+
+Based on the `PluginCacheKeyType` definitions (Image, Registry, Global), here’s a breakdown of how cache keys would be generated:
+
+1. `GlobalPluginCacheKeyType` (Global)
+   - Global cache key type will not be allowed to use with the `tokenAttributes.serviceAccountTokenAudience` field configured for the plugin.
+2. `RegistryPluginCacheKeyType` (Registry)
+   - The cache key will be generated based on the image registry URL, service account metadata (namespace, name, UID), and hash of the service account annotations that are passed to the plugin in the `CredentialProviderRequest.ServiceAccountAnnotations` field.
+3. `ImagePluginCacheKeyType` (Image)
+    - The cache key will be generated based on the image URL, service account metadata (namespace, name, UID), and hash of the service account annotations that are passed to the plugin in the `CredentialProviderRequest.ServiceAccountAnnotations` field.
+
+#### How will this work with the Ensure secret pull images KEP?
+
+[2535-ensure-secret-pull-images](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images) is a KEP that aims to give admin the ability to ensure pods that use a image are authorized to access the image. The KEP doesn't factor in the use of service account tokens for image pull. As part of the alpha implementation the 2 features together will not provide a desired outcome. Post alpha, we'll need to update the Ensure secret pull images KEP to factor in the use of service account tokens for image pull.
+
+Notes from reviewing the current KEP and discussion with @stalz:
+
+1. Different KSA for same image should result in image pull from registry.
+2. Same KSA for the image will result in allowing pod to use the image.
+   1. How would expiry work in this scenario?
+      1. I think it'll just be tied to the KSA and not so much to the expiry of the token because we're doing the same thing with image pull secrets (considered valid until deleted and recreated). Any change to KSA will result in change in UID at the minimum and that'll result in KSA not found in cache for the image (assuming the key used to store in cache is consistent with the cache key used in the credential provider cache that takes UID into consideration). Need to share the cache key generation logic to be consistent.
```

Comment:
A change to a UID of an SA means it got deleted and recreated, thus making it a completely different SA, but simple updates to an SA won't cause it to change UIDs. The UID in the cache key protects the system from attempts on name clashes.

Reply:
Deletion and recreation of SA with the same name in the namespace is what I meant here, I'll clarify it.